src/fpm_compiler.f90

```diff
@@ -416,12 +416,79 @@ subroutine set_preprocessor_flags (id, flags)
     do i = 1, size(package%preprocess)
         if (package%preprocess(i)%name == "cpp") then
             flags = flag_cpp_preprocessor// flags
+            call get_macros_from_manifest(id, flags, i)
             exit
         end if
     end do
 
 end subroutine set_preprocessor_flags
 
+!> This subroutine will check and get all the macros that are defined in the manifest file
+!> under preprocess section for a specific preprocessor
+subroutine get_macros_from_manifest(id, flags, index_of_preprocessor)
+    integer(compiler_enum), intent(in) :: id
+    type(package_config_t) :: package
+    type(error_t), allocatable :: error
+    character(len=:), allocatable :: flags
+    character(len=:), allocatable :: macro_definition_symbol
+    character(:), allocatable :: valued_macros(:)
+    character(len=:), allocatable :: version
+
+    integer :: i
+    integer :: index_of_preprocessor
+
+    call get_package_data(package, "fpm.toml", error)
```

Comment:
At the same time, the `package` parameter in `set_preprocessor_flags` can also be passed in from `fpm.f90::build_model::Line81`.
(This may avoid opening and closing `fpm.toml` multiple times)